### PR TITLE
MULE-19537: App get stuck when is trying to read the error payload in the error handler after a failing http request

### PR DESCRIPTION
**Pull request by rbourbonnavarro**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/privileged/processor/chain/AbstractMessageProcessorChain.java [hunk 1]
```diff
@@ -227,6 +235,17 @@ public CoreEvent process(CoreEvent event) throws MuleException {
     }
   }
 
+  private void routeError(Consumer<Exception> errorRouter, Exception throwable) {
+    if (!isTransactionActive() && schedulerService.isCurrentThreadSwitchOnErrorRequired()) {
+      if (errorSwitchScheduler == null) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/privileged/processor/chain/AbstractMessageProcessorChain.java:240]: how could this be null? this should be done in initialization

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mule/runtime/core/privileged/processor/chain/AbstractMessageProcessorChain.java (hunk 1 above)
Comment (rbourbonnavarro) [reply core/src/main/java/org/mule/runtime/core/privileged/processor/chain/AbstractMessageProcessorChain.java]: Moved to the `initialise` phase

**Comment by rbourbonnavarro** [comment]
--validate

File: core-tests/src/test/java/org/mule/runtime/core/internal/exception/AbstractErrorHandlerTestCase.java [hunk 2]
```diff
@@ -89,7 +88,7 @@ public void before() throws Exception {
 
   @After
   public void after() {
-    flow.dispose();
+    // Nothing to do
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/exception/AbstractErrorHandlerTestCase.java:91]: just remove the method

File: core-tests/src/test/java/org/mule/runtime/core/privileged/processor/MessageProcessorsTestCase.java [hunk 3]
```diff
@@ -150,7 +152,9 @@ public void processToApplyMap() throws Exception {
 
   @Test
   public void processToApplyMapInChain() throws Exception {
-    assertThat(processToApply(input, createChain(map)), is(output));
+    Processor chain = createChain(map);
+    assertThat(processToApply(input, chain), is(output));
+    disposeChain(chain);
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/privileged/processor/MessageProcessorsTestCase.java:157]: make this in a finally block

Comment (elrodro83) [reply core-tests/src/test/java/org/mule/runtime/core/privileged/processor/MessageProcessorsTestCase.java:157]: or make chain into an attribute and dispose it an `@After` method

File: core-tests/src/test/java/org/mule/runtime/core/privileged/processor/chain/DefaultMessageProcessorChainTestCase.java [hunk 4]
```diff
@@ -188,13 +190,21 @@ public void before() throws MuleException {
...
+    }
   }
 
   @Test
   public void testMPChain() throws Exception {
     DefaultMessageProcessorChainBuilder builder = new DefaultMessageProcessorChainBuilder();
     builder.chain(getAppendingMP("1"), getAppendingMP("2"), getAppendingMP("3"));
-    assertEquals("0123", process(builder.build(), getTestEventUsingFlow("0")).getMessage().getPayload().getValue());
+    messageProcessor = builder.build();
+    assertEquals("0123", process(messageProcessor, getTestEventUsingFlow("0")).getMessage().getPayload().getValue());
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/privileged/processor/chain/DefaultMessageProcessorChainTestCase.java:207]: do you feel like migrating these assertions to hamcrest while you are at it?

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/nb/NonBlockingOperationsTestCase.java [hunk 5]
```diff
@@ -51,6 +68,16 @@ protected String getConfigFile() {
     return "iron-man-config.xml";
   }
 
+  @Before
+  public void setUp() {
+    ThreadCaptor.setCapturedThreads(newKeySet());
```
Comment (elrodro83) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/nb/NonBlockingOperationsTestCase.java:73]: static import

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/nb/NonBlockingOperationsTestCase.java [hunk 6]
```diff
@@ -51,6 +68,16 @@ protected String getConfigFile() {
...
+  }
+
+  @After
+  public void tearDown() {
+    ThreadCaptor.setCapturedThreads(null);
```
Comment (elrodro83) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/nb/NonBlockingOperationsTestCase.java:78]: static import

File: modules/extensions-spring-support/src/test/resources/iron-man-config.xml [hunk 7]
```diff
@@ -30,8 +30,18 @@
         <error-handler>
             <on-error-propagate>
                 <logger/>
-                <!-- TODO: MULE-18118 - uncomment the following line when this issue is fixed -->
-                <!-- <flow-ref name="captureThread"/> -->
+                <flow-ref name="captureThread"/>
```
Comment (elrodro83) [inline modules/extensions-spring-support/src/test/resources/iron-man-config.xml:33]: 🥳

**Review by elrodro83 — DISMISSED** [review]

File: core-tests/src/test/java/org/mule/runtime/core/privileged/processor/chain/DefaultMessageProcessorChainTestCase.java (hunk 4 above)
Comment (rbourbonnavarro) [reply core-tests/src/test/java/org/mule/runtime/core/privileged/processor/chain/DefaultMessageProcessorChainTestCase.java]: Done.

File: core-tests/src/test/java/org/mule/runtime/core/privileged/processor/MessageProcessorsTestCase.java (hunk 3 above)
Comment (rbourbonnavarro) [reply core-tests/src/test/java/org/mule/runtime/core/privileged/processor/MessageProcessorsTestCase.java]: Done.

**Review by fsgonz — APPROVED** [review]

File: core-tests/src/test/java/org/mule/runtime/core/internal/exception/AbstractErrorHandlerTestCase.java (hunk 2 above)
Comment (rbourbonnavarro) [reply core-tests/src/test/java/org/mule/runtime/core/internal/exception/AbstractErrorHandlerTestCase.java]: Done.

**Comment by rbourbonnavarro** [comment]
--validate

**Review by fsgonz — APPROVED** [review]

**Pull request merged**